### PR TITLE
Update index.qmd

### PR DESCRIPTION
**Pull request by willtebbutt**
Addresses part of #547 .

@gdalle does this read more correctly?

### CODE REVIEW

File: tutorials/docs-10-using-turing-autodiff/index.qmd
Comment (willtebbutt) [inline tutorials/docs-10-using-turing-autodiff/index.qmd:23]:
```suggestion
For `ReverseDiff`, pass `adtype=AutoReverseDiff()` to the sampler constructor. An additional argument can be provided to `AutoReverseDiff` to specify whether to to cache the tape only once and reuse it later use (`false` by default, which means no caching). This can substantially improve performance, but risks silently incorrect results if not used with care.

```

**Comment by gdalle** [comment]
I think there might still be a misunderstanding here. The keyword argument `AutoReverseDiff(; compile)` only controls whether the tape is optimized after being recorded. It doesn't control whether a recorded tape is only used once, or reused across function calls (inducing possible control flow problems).
Tape reuse is instead determined by how the `ADGradient` is created, and whether or not an example vector `x` is passed to it. See these key lines in LogDensityProblemsAD.jl:

https://github.com/tpapp/LogDensityProblemsAD.jl/blob/e3401f21b5a065df0d5de38b37fad0e6650618f3/ext/LogDensityProblemsADReverseDiffExt.jl#L45-L56

So the real question is:

- in which places does Turing "prepare" the `ADGradient` with an example `x`
- can the user of the library do anything about it, if they know that their loglikelihood has value-dependent control flow?

**Comment by torfjelde** [comment]
> in which places does Turing "prepare" the ADGradient with an example x

Typically this occurs in the initial step of the sampler, i.e. once (there are exceptions, but in those cases you can't really do much better in our case).

> can the user of the library do anything about it, if they know that their loglikelihood has value-dependent control flow?

Unfortunately not for ReverseDiff.jl in compiled mode (they can however just not use compiled mode).

File: tutorials/docs-10-using-turing-autodiff/index.qmd
Comment (torfjelde) [inline tutorials/docs-10-using-turing-autodiff/index.qmd:28]:
```suggestion
For instance, `if`-statements with conditions that can be determined at compile time or conditions that depend only on fixed properties of the model, e.g. fixed data.
```

I don't think there's much point mentioning minibatching here, as there's no "easy to use" support for this in Turing.jl and so it's really not something people do much of (don't think I've ever seen anyone actually do this in applications with Turing.jl).

File: tutorials/docs-10-using-turing-autodiff/index.qmd
Comment (torfjelde) [inline tutorials/docs-10-using-turing-autodiff/index.qmd:22]:
```suggestion
For `ForwardDiff`, pass `adtype=AutoForwardDiff(; chunksize)` to the sampler constructor. A `chunksize` of `nothing` permits the chunk size to be automatically determined. For more information regarding the selection of `chunksize`, please refer to [related section of `ForwardDiff`'s documentation](https://juliadiff.org/ForwardDiff.jl/dev/user/advanced/#Configuring-Chunk-Size).

```

Comment (torfjelde) [reply tutorials/docs-10-using-turing-autodiff/index.qmd:22]: Maybe also add this, as @gdalle pointed out, this is not actually the case anymore (this was a left-over thing from using LogDensityProblemsAD.jl I believe, but this changed when it all moved to ADTypes.jl backed)

**Review by torfjelde — COMMENTED** [review]
Added some comments 👍

**Comment by gdalle** [comment]
> Unfortunately not for ReverseDiff.jl in compiled mode (they can however just not use compiled mode).

@torfjelde I think that's the key confusion between us. My understanding is that, _regardless of whether it is compiled or not_, the tape only captures execution for one version of the control flow. Thus, the tape always becomes invalid if the control flow changes (e.g. different branch of an `if` statement). Isn't that how it goes?

Source: <https://juliadiff.org/ReverseDiff.jl/dev/api/#The-AbstractTape-API>

**Comment by devmotion** [comment]
The tape is purely internal and neither stored nor reused if `compile = Val(false)`.

> Tape reuse is instead determined by how the ADGradient is created, and whether or not an example vector x is passed to it

This is wrong. The tape is only reused if `compile = Val(true)`: https://github.com/tpapp/LogDensityProblemsAD.jl/blob/e3401f21b5a065df0d5de38b37fad0e6650618f3/ext/LogDensityProblemsADReverseDiffExt.jl#L51-L56

**Comment by gdalle** [comment]
Okay then IIUC it's a matter of inconsistency between
- the "compile" kwarg here and in LogDensityProblemsAD (which determines whether a tape is reused)
- the compilation as defined by ReverseDiff, which is about speeding up an existing tape (and disconnected from the reuse issue)

**Comment by gdalle** [comment]
Note that in ADTypes, the "compile" argument to AutoReverseDiff is defined ambiguously too. So we should perhaps add more details to the struct, something like AutoReverseDiff(tape=true, compile=false)?

**Comment by gdalle** [comment]
I'd love for you to chime in here @devmotion @torfjelde: https://github.com/SciML/ADTypes.jl/issues/91

**Comment by gdalle** [comment]
> I'm not too opinionated about this, as compilation without caching seems somewhat useless? Are there scenarios where you'd like to do that?

Indeed you can't compile a tape you never record in the first place. In any case, I think the ambiguous terminology was fixed by https://github.com/SciML/ADTypes.jl/issues/91. It's just a shame that the word "compile" was chosen instead of "record", given how both are used in ReverseDiff's documentation. But it's a sunk cost now.

File: tutorials/docs-10-using-turing-autodiff/index.qmd
Comment (gdalle) [inline tutorials/docs-10-using-turing-autodiff/index.qmd:23]:
```suggestion
For `ReverseDiff`, pass `adtype=AutoReverseDiff()` to the sampler constructor. An additional keyword argument called `compile` can be provided to `AutoReverseDiff`. It specifies whether to pre-record the tape only once and reuse it later (`compile` is set to `false` by default, which means no pre-recording). This can substantially improve performance, but risks silently incorrect results if not used with care.

```

File: tutorials/docs-10-using-turing-autodiff/index.qmd
Comment (gdalle) [inline tutorials/docs-10-using-turing-autodiff/index.qmd:26]:
```suggestion
Pre-recorded tapes should only be used if you are absolutely certain that the sequence of operations performed in your code does not change between different executions of your model.

```

**Comment by willtebbutt** [comment]
Thank you all for raising this problem / helping to fix it -- it's good that we've gotten to the bottom of this! I'm basically happy with this. @gdalle and @torfjelde could take a final look at this, and let me know whether you are happy.

**Comment by torfjelde** [comment]
> It's just a shame that the word "compile" was chosen instead of "record", given how both are used in ReverseDiff's documentation. But it's a sunk cost now.

I agree with that 😕

**Review by gdalle — APPROVED** [review]

**Review by torfjelde — APPROVED** [review]
LGTM! Added one final comment that would be nice to shoehorn in here since we're already making alterations, but looks good:)

File: tutorials/docs-10-using-turing-autodiff/index.qmd
Comment (gdalle) [reply tutorials/docs-10-using-turing-autodiff/index.qmd:69]: When you fully switch to DI you can also use `DifferentiationInterfaceTest.benchmark_differentiation` for this ;)

Comment (willtebbutt) [reply tutorials/docs-10-using-turing-autodiff/index.qmd:69]: This is a good point. @torfjelde would it make more sense to just link to TuringBenchmarking.jl, rather than copying + pasting in a code snippet which may become outdated? Alternatively, is there a simple way to put example benchmarking code in here which runs whenever the docs are built?

Comment (willtebbutt) [reply tutorials/docs-10-using-turing-autodiff/index.qmd]: Thanks for reminding me about this @torfjelde . I've made the changes requested, so I'm going to mark this conversation closed. I'll get CI passing and then merge.

**Comment by willtebbutt** [comment]
@penelopeysm it looks like my PR has the potential to cause merge conflicts with #559 . Would you like me to wait until you've merged yours before merging this, or are you happy to resolve any conflicts that arise?

**Comment by penelopeysm** [comment]
I'm happy to sort it out :)

**Pull request merged**
